tensorflow/core/kernels/conv_ops_gpu_3.cu.cc

```diff
@@ -413,6 +429,275 @@ struct PadInput<GPUDevice, T, int, NDIMS> {
+// number of threads per SM limit, shared memory limit and overall degree of
+// parallelism.
+// Tile size combinations lying on this frontier would achieve the maximum
+// utilization of available resources and combinations lying outside this
+// frontier is not performance-wise profitable or simply not possible.
+#define TPF_CHECKER(OP)                                                   \
+  size_of_t == 16 &&    (TileLongSide == 32   && (TileShortSide OP 4)  || \
+                         TileLongSide == 64   && (TileShortSide OP 4)  || \
+                         TileLongSide == 128  && (TileShortSide OP 4)  || \
+                         TileLongSide == 256  && (TileShortSide OP 2)) || \
+      size_of_t == 8 && (TileLongSide == 32   && (TileShortSide OP 15) || \
+                         TileLongSide == 64   && (TileShortSide OP 15) || \
+                         TileLongSide == 128  && (TileShortSide OP 8)  || \
+                         TileLongSide == 256  && (TileShortSide OP 4)  || \
+                         TileLongSide == 512  && (TileShortSide OP 2)) || \
+      size_of_t == 4 && (TileLongSide == 32   && (TileShortSide OP 15) || \
+                         TileLongSide == 64   && (TileShortSide OP 15) || \
+                         TileLongSide == 128  && (TileShortSide OP 15) || \
+                         TileLongSide == 256  && (TileShortSide OP 10) || \
+                         TileLongSide == 512  && (TileShortSide OP 4)  || \
+                         TileLongSide == 1024 && (TileShortSide OP 2)) || \
+      size_of_t == 2 && (TileLongSide == 32   && (TileShortSide OP 15) || \
+                         TileLongSide == 64   && (TileShortSide OP 15) || \
+                         TileLongSide == 128  && (TileShortSide OP 15) || \
+                         TileLongSide == 256  && (TileShortSide OP 10) || \
+                         TileLongSide == 512  && (TileShortSide OP 4)  || \
+                         TileLongSide == 1024 && (TileShortSide OP 2)) || \
+      size_of_t == 1 && (TileLongSide == 32   && (TileShortSide OP 15) || \
+                         TileLongSide == 64   && (TileShortSide OP 15) || \
+                         TileLongSide == 128  && (TileShortSide OP 15) || \
+                         TileLongSide == 256  && (TileShortSide OP 10) || \
+                         TileLongSide == 512  && (TileShortSide OP 4)  || \
+                         TileLongSide == 1024 && (TileShortSide OP 2))
+
+constexpr bool TileSizePossibilityFrontierCheck(int TileLongSide,
+                                                int TileShortSide,
+                                                int size_of_t, int mode) {
+  // When mode is 0, this function checks whether the tile size combination lies
+  // on the tile size possibility frontier. When mode is 1, it checks whether
+  // the combination lies outside the frontier.
+  return mode == 0 ? TPF_CHECKER(==) : TPF_CHECKER(>);
+}
```

Comment:
I am kind of tempted to suggest that we get rid of this macro too:

```
// We want std::equal_to and std::greater, but they're not constexpr until C++14.
struct EqualTo {
  constexpr bool operator()(int a, int b) { return a == b; }
};
struct GreaterThan {
  constexpr bool operator()(int a, int b) { return a > b; }
};
template <typename Op>
constexpr bool TileSizeFrontierCheck(..., Op op) {
  return size_of_t == 16 && ((TileLongSide == 32 && op(TileShortSide, 4)) ||
                             (TileLongSide == 64 && op(TIleShortSide, 4)) || ...
                                            
}
constexpr bool TileSizeOnFrontier(...) { return TileSizeFrontierCheck(..., EqualTo()); }
constexpr bool TileSizePastFrontier(...) { return TileSizeFrontierCheck(..., GreaterThan()); }
```

In addition to getting rid of the macro, this also lets you avoid the sort of ugly `mode` param.

I also renamed it "tile size frontier", which I think gets the point across just as well as "tile size possibility frontier" with one less word.

WDYT?